packages/library/src/commonMain/kotlin/io/realm/Realm.kt

```diff
@@ -134,9 +140,23 @@ class Realm private constructor(configuration: RealmConfiguration, dbPointer: Na
     private suspend fun updateRealmPointer(newRealm: NativePointer, newVersion: VersionId) {
         realmPointerMutex.withLock {
             log.debug("Updating Realm version: $version -> $newVersion")
-            if (newVersion >= version) {
-                realmReference = RealmReference(this, newRealm)
+            val newReference = RealmReference(this, newRealm)
+            val intermediateReference = if (newVersion >= version) {
+                val version = Pair(realmReference.dbPointer, WeakReference(realmReference))
+                realmReference = newReference
+                version
```

Comment:
You are returning `version` here and a Pair in the else. This seems wrong?

Reply:
Also the naming in this section is a bit hard to follow, we have`newReference`, `RealmReference`,  `WeakReference`, `intermediateReference` and `intermediateReferences`, at least it makes my head spin a little when trying to reason about it. Not exactly sure I have a better proposal though.

Reply:
The name was wrong, so it was actually returning a reference. I tried to separate the _reference tracking_ into a method to hide building the pairs, weak references, etc. Hopefully it is a bit more clear now.